---
title: "Exploring with PRSs and ROS/MAP"
output: github_document
---

## Packages and importing datasets:

```{r setup, include=FALSE}
library(data.table)
library(openxlsx)
library(rms)
library(limma)
library(ggplot2)
library(factoextra)
library(diptest)
library(corrplot)

ROSmaster <- readRDS("C:/Users/amink/OneDrive/Documents/Current Jobs/Masters Thesis/Code/Datasets/ROSMAP_Phenotype/ROSmaster.rds")

meta_pheno <- read.csv("C:/Users/amink/OneDrive/Documents/Current Jobs/Masters Thesis/Code/Datasets/Sum_Stats/Pan_UK_Biobank_phenotype_manifest_h2_more_0.05_both_sex_non_ambigous.csv")

geno_pcs <- read.table("/Users/amink/OneDrive/Documents/Current Jobs/Masters Thesis/Code/Datasets/PCA_Genotype/geno_qc.eigenvec.txt",header=F)
names(geno_pcs) <- c("FID","IID",paste0("genoPC",seq(1:10)))

prs_filenames <- list.files("/Users/amink/OneDrive/Documents/Current Jobs/Masters Thesis/Code/Datasets/PRS_PLINK/",full.names = T)

# read in PRS files
prs_list <- list()
for (prsfile in prs_filenames[1:10]) {
  prs_list[[tail(strsplit(prsfile,split="/")[[1]],n=1)]] <- as.data.frame(fread(prsfile,header=T,sep=","))
}
```

## LOOP for PCA analysis

```{r setup, include=FALSE}
dsets_torun <- names(prs_list)

results1 <- list()
for (thresh in dsets_torun) {
  pval <- gsub(".txt","",thresh)
  print(paste("Starting run, p-value threshold=",pval))
  print(paste("number of subjects =",nrow(prs_list[[thresh]])))
  print(paste("number of PRS =",ncol(prs_list[[thresh]])-2))
  
  formod <- merge(prs_list[[thresh]],geno_pcs,by=c("IID","FID"))

  # this could be removed from loop, but in case any subjects are removed/added to either geno_pcs or PRS files, the design matrix will need to be recalculated, so left in to prevent error
  designvars <- c(names(geno_pcs)[-c(1:2)])  
  designtext <- paste("model.matrix( ~ ",paste("formod$",designvars,sep="",collapse=" + "),")",sep="")
  design <- eval(parse(text=designtext))

  # OLS regression using machinery build for gene expression in limma
  print(paste("Fitting model with genetic PCs"))
  fit <- lmFit(t(formod[,3:ncol(prs_list[[thresh]])]),design)
  
  print(paste("Extracting model residuals"))
  res <- t(residuals.MArrayLM(fit,y = t(formod[,3:ncol(prs_list[[thresh]])])))
  res <- apply(res,2,scale)

  # flag multimodal scores
  print(paste("Flagging scores with multimodal distributions"))
  dts <- apply(res,2,dip.test)
  dtsp <- lapply(dts,function(x) { x$p.value })
  multimodal <- which(dtsp < 0.05/ncol(res))
  
  print(paste("Removing",length(multimodal),"multimodal scores"))
  
  if (length(multimodal)>0) {
    multimodal.names <- colnames(res)[multimodal]
    res.clean <- res[,-multimodal]
  } else {
    multimodal.names <- NA
    res.clean <- res
  }
  
  # run PCA
  print(paste("Running PCA"))
  respca <- prcomp(res.clean)
  
  # Calculate subject and variable contributions to each component
  print(paste("Calculating variable and subject contributions to components"))
  res.var <- get_pca_var(respca)
  res.ind <- get_pca_ind(respca)
  
  # Save files in list
  print(paste("Saving data"))
  res.clean <- as.data.frame(res.clean)
  res.clean$IID <- formod$IID
  allres <- list(residuals=res.clean,
                 multimodal=multimodal.names,
                 pca=respca,
                 res.var=res.var,
                 res.ind=res.ind)
  
  saveRDS(allres,file=paste0("PCAresults_",pval,".rds"))
  results1[[pval]] <- allres
}
```


```{r setup, include=FALSE}
prs_matrix_0.0001 <- as.data.frame(results1$p_val_0_0001$residuals[,-1158])
names(prs_matrix_0.0001) <- c(paste0('V',c(1:(dim(prs_matrix_0.0001)[2]))))
clusters <- hclust(dist(t(prs_matrix_0.0001)))
plot(clusters, col='lightblue',xlab=' ', main =paste0('Dendogram of PRS of ',
                                                      dim(prs_matrix_0.0001)[2],' phenotypes'))
                                                      
                                                      
clusters1 <- hclust(dist(prs_matrix_0.0001))
plot(clusters1, col='lightblue',xlab=' ', main =paste0('Dendogram of PRS of individuals of ',
                                                      dim(prs_matrix_0.0001)[1],' individuals'))


fviz_eig(results1$p_val_0_0001$pca,ncp = 10)
#rownames(results1$p_val_0_05$res.var$cos2)[order(results1$p_val_0_05$res.var$cos2[,8])][1:10]
heatmap(as.matrix(results1$p_val_0_0001$residuals[,-1158]))
``` 
## flag scores with multimodal distributions

# visualize PCA
```{r setup, include=FALSE}
# merge with phenotype data


res.pcs <- as.data.frame(results1$p_val_1e_05$pca$x)
res.pcs$IID <- results1$p_val_1e_05$residuals$IID

pheno_pcs <- merge(res.pcs,ROSmaster,by="IID")
pheno_pcs2 <- merge(pheno_pcs,geno_pcs,by="IID")

ggplot(data=pheno_pcs2,aes(x=PC8,y=cogn_global_random_slope))+
  geom_point()+
  #geom_point(data=subset(pheno_pcs2,IID %in% highlightiid),col="red",size=3)+
  geom_smooth(method="lm")+
  theme_minimal()


## identify relationships of PCs with phenotypes of interest
pheno.list <- c("amyloid_sqrt","tangles_sqrt","cogn_global_random_slope","age_death","parksc_lv_sqrt","neuroticism_12","anxiety_20items","cesdsum_lv","educ","mf3123","it3123","vm3123","pput3123")

PClist <- paste0("PC",seq(1:100))

index <- 1
pvalues <- NULL
bvalues <- NULL
nvalues <- NULL
phenovalues <- NULL
pcvalues <- NULL
for (pheno in pheno.list) {
  for (PC in PClist) {
    form <- formula(paste(pheno,"~",PC))
    mod <- lm(data=pheno_pcs2,form)
    pvalues[index] <- anova(mod)[1,5]
    bvalues[index] <- coef(mod)[2]
    nvalues[index] <- dim(mod$model)[1]
    pcvalues[index] <- PC
    phenovalues[index] <- pheno
    index <- index + 1
   }
}

assocres <- data.frame(pheno=phenovalues,
                       pc=pcvalues,
                       b=bvalues,
                       p=pvalues,
                       n=nvalues,
                       fdr=p.adjust(pvalues))

ggplot(data=assocres, aes(x=pheno,y=-log10(p),group=pc))+
  geom_bar(stat="identity",position="dodge")+
  geom_hline(yintercept = -log10(0.05/nrow(assocres)),col="red",lty=2)+ 
  geom_text(data=subset(assocres,p<0.05/nrow(assocres)),aes(label=pc))+
  theme_minimal()+
  theme(axis.text.x=element_text(angle = -45, hjust = 0))
  

### get contributing PRS for given PC
pcnum <- 25
results1$p_val_1e_05$res.var$cos2[,pcnum][order(results1$p_val_1e_05$res.var$cos2[,pcnum],decreasing = T)][1:10]

### get cos2 for top 5 components

corrplot(results1$p_val_1e_05$res.var$cos2[1:10,1:5],is.corr = F,tl.cex = 0.3)

fviz_cos2(results1$p_val_1e_05$pca,choice="var",axes=8, top=10)
fviz_contrib(results1$p_val_1e_05$pca,choice="var",axes=25, top=30)


pheno_pcs3 <- merge(pheno_pcs2,prs_list$p_val_1e_05.txt,by="IID")
summary(lm(data=pheno_pcs3, cogn_global_random_slope ~ PC9))
summary(lm(data=pheno_pcs3, cogn_global_random_slope ~ icd10_C71_C71_Malignant_neoplasm_of_brain))



##########################
pc25list <- names(results1$p_val_1e_05$res.var$cos2[,25][order(results1$p_val_1e_05$res.var$cos2[,25],decreasing = T)][1:20])
pc7list <- names(results1$p_val_1e_05$res.var$cos2[,7][order(results1$p_val_1e_05$res.var$cos2[,7],decreasing = T)][1:20])
pc8list <- names(results1$p_val_1e_05$res.var$cos2[,8][order(results1$p_val_1e_05$res.var$cos2[,8],decreasing = T)][1:10])

pheno.list <- c("amyloid_sqrt","tangles_sqrt","cogn_global_random_slope")

PClist <- c(pc8list,"PC8")

index <- 1
pvalues <- NULL
bvalues <- NULL
nvalues <- NULL
phenovalues <- NULL
pcvalues <- NULL
rvalues <- NULL
for (pheno in pheno.list) {
  for (PC in PClist) {
    form <- formula(paste0(pheno," ~ `",PC,"`"))
    mod <- lm(data=pheno_pcs3,form)
    summod <- summary(mod)
    rvalues[index] <- summod$r.squared
    pvalues[index] <- anova(mod)[1,5]
    bvalues[index] <- coef(mod)[2]
    nvalues[index] <- dim(mod$model)[1]
    pcvalues[index] <- PC
    phenovalues[index] <- pheno
    index <- index + 1
  }
}

assocres25 <- data.frame(pheno=phenovalues,
                       pc=pcvalues,
                       b=bvalues,
                       p=pvalues,
                       n=nvalues,
                       r=rvalues,
                       fdr=p.adjust(pvalues))


```